### ISSUE TITLE
Issues finding d3 when bundled via browserify

### ISSUE THREAD

**Issue by mshwery** [novus/nvd3#1505]
Receiving this error because d3 is not attached to the window object: `d3 is not defined`

I'm hitting a snag trying out this library in conjunction with browserify/babel. Is browserify not supported? Why does nvd3 depend on d3 being a global?

There a few moving parts at play, but here's my basic setup:

package.json
```js
  "browserify-shim": {
    "angular": {
      "exports": "angular"
    },
    "jquery": "$",
    "d3": "d3",
    "nvd3": {
      "depends": [
        "d3"
      ],
      "exports": "nv"
    },
    "restangular": {
      "depends": [
        "angular",
        "lodash:_"
      ]
    }
  }
```

I've got plenty of other shimmed libs that work fine, including dependencies. Browserify is creating a bundle of all my vendor libs and one for all my other code. This is working as expected. The only noteable difference between how the two bundles are generated (again works before nvd3 fine):

```js
var vendorBundle = browserify({ debug: false });

var appBundle = browserify({
    entries: [config.srcDirectory + '/js/app.js'],
    debug: debug,
    noParse: dependencies
  })
  .transform(html)
  .transform(babelify.configure({
    optional: ['runtime'],
    ignore: /(.*\.html|.*node_modules\/*.*)/
  }));
```

in use:
```js
import d3 from 'd3';
import nv from 'nvd3'; // because i've imported nvd3, nvd3 gets evaluated and _breaks_
```

**Comment by liquidpele** [comment]
I don't know anything about browserify sorry... if you can figure out a way to make it compatible then send us a pull request!

**Comment by mshwery** [comment]
Ah I see the problem – it's definitely that [d3 is expected to be a _global_ variable](https://github.com/novus/nvd3/blob/f6c337cfd18b57ba89ccff850b8bc7ce4eafc223/build/nv.d3.js#L16). This is problematic for all module bundler consumers which do not expose packages as globals.

The latest release [doesn't support CommonJS](https://github.com/novus/nvd3/blob/f6c337cfd18b57ba89ccff850b8bc7ce4eafc223/build/nv.d3.js#L16), but I see that the [current version of `master` does](https://github.com/novus/nvd3/blob/master/build/nv.d3.js#L16-L19)

I would encourage this lib to support both vanilla js consumers and module system users by providing a [UMD](https://github.com/umdjs/umd) wrapper so that RequireJS users, Node/CommonJS users and webpack/browserify users can all consume this library easily.

I think as long as that latest version of master get released I can use it with browserify + browserify-shim.

**Comment by mshwery** [comment]
@liquidpele just curious when another release might happen, since that would resolve this issue.

**Comment by thepianist2** [comment]
i have this same problem.

**Comment by liquidpele** [comment]
Probably won't be for a month or so.

**Comment by Zeychiel** [comment]
```
var d3 = require('d3');
window.d3 = d3;
require('nvd3')
```
will solve this, waiting for the master to be released.
Yes, it's ugly :D

**Comment by arvindkumarc** [comment]
I still face this issue when using along with browserify.

**Comment by arvindkumarc** [comment]
There is a fix provided for `require` https://github.com/novus/nvd3/commit/b474a441ec340f87e5c1a3f2f9339ebf4516500b#diff-7eb52b366866677666470e019283c8ea

This is not part of the build `nv.d3.js`

**Comment by gschmottlach-xse** [comment]
That fix above (which seems to be checked into master) does **not** fix it for my use-case.

```
// Node/CommonJS - require D3
if (typeof(module) !== 'undefined' && typeof(exports) !== 'undefined' && typeof(d3) == 'undefined') {
    d3 = require('d3');
}
```

I'm using this module in an Electron application. In my environment (Electron1.4.5 and Node 6.5.0) 'd3' remains undefined because although 'module' is defined, 'exports' is **not**. So I am wondering if this check is a little too restrictive? Is it sufficient just to verify that 'd3' is defined, and if not, do the d3 = require('d3')? Perhaps something like:

```
// Node/CommonJS - require D3
if ( typeof(d3) == 'undefined' ) {
    var d3 = require('d3');
}
```

Ultimately I'm using the angular-nvd3 module (https://krispo.github.io/angular-nvd3/#/). It's broken due to the reason above and also due to the fact that 1.7.1 has a later date than the 1.8.4 release. As a result npm continues to pull 1.7.1 rather than 1.8.4 as the dependency. If the above change could be made and a new release cut (1.8.5) it would resolve both issues.  Is this possible? Is my request breaking another use case?
